src/covid_model_seiir_pipeline/forecasting/postprocessing_lib.py

```diff
@@ -0,0 +1,181 @@
+        data_interface=data_interface,
+    )
+    draws = range(data_interface.get_n_draws())
+    with multiprocessing.Pool(FORECAST_SCALING_CORES) as pool:
+        beta_residuals = pool.map(_runner, draws)
+    return beta_residuals
+
+
+def load_beta_residuals_by_draw(draw_id: int, data_interface: ForecastDataInterface) -> pd.Series:
+    beta_regression = data_interface.load_beta_regression(draw_id)
+    beta_regression = (beta_regression
+                       .set_index(['location_id', 'date'])
+                       .sort_index()[['beta', 'beta_pred']])
+    beta_residual = np.log(beta_regression['beta'] / beta_regression['beta_pred']).rename(draw_id)
+    return beta_residual
+
+
+def concat_draws(measure_data: List[pd.Series]) -> pd.DataFrame:
+    # 3x faster than pd.concat for reasons I don't understand.
+    measure_data = functools.reduce(lambda a, b: pd.merge(a, b, left_index=True, right_index=True, how='outer'),
+                                    measure_data)
+    measure_data.columns = [f'draw_{i}' for i in measure_data.columns]
+    return measure_data.reset_index()
+
+
+def build_resampling_map(deaths, resampling_params):
+    cumulative_deaths = deaths.groupby(level='location_id').cumsum()
+    max_deaths = cumulative_deaths.groupby(level='location_id').max()
+    upper_deaths = max_deaths.quantile(resampling_params['upper_quantile'], axis=1)
+    lower_deaths = max_deaths.quantile(resampling_params['lower_quantile'], axis=1)
+    resample_map = {}
+    for location_id in max_deaths.index:
+        upper, lower = upper_deaths.at[location_id], lower_deaths.at[location_id]
+        loc_deaths = max_deaths.loc[location_id]
+        to_resample = loc_deaths[(upper < loc_deaths) | (loc_deaths < lower)].index.tolist()
+        np.random.seed(12345)
+        to_fill = np.random.choice(loc_deaths.index, len(to_resample), replace=False).tolist()
+        resample_map[location_id] = {'to_resample': to_resample,
+                                     'to_fill': to_fill}
+    return resample_map
+
+
+def summarize(data: pd.DataFrame):
+    data['mean'] = data.mean(axis=1)
+    data['upper'] = data.quantile(.975, axis=1)
+    data['lower'] = data.quantile(.025, axis=1)
+    return data[['mean', 'upper', 'lower']]
+
+
+def resample_draws(resampling_map, *measure_data: pd.DataFrame):
+    _runner = functools.partial(
+        resample_draws_by_measure,
+        resampling_map=resampling_map
+    )
+    with multiprocessing.Pool(FORECAST_SCALING_CORES) as pool:
+        resampled = pool.map(_runner, measure_data)
+    return resampled
+
+
+def resample_draws_by_measure(measure_data: pd.DataFrame, resampling_map):
+    output = []
+    for location_id, loc_map in resampling_map.items():
+        loc_data = measure_data.loc[location_id]
+        loc_data[loc_map['to_resample']] = loc_data[loc_map['to_fill']]
+        loc_data = pd.concat({location_id: loc_data}, names=['location_id'])
+        if isinstance(loc_data, pd.Series):
```

Comment:
What causes this check to need to exist?

Reply:
Pandas being infuriating and not giving me the thing I want.

Reply:
If you call `df.loc[single_value_indexer]` on a dataframe with a multi-index, it looks in the outermost level of the index, plucks out the sub_df corresponding to the indexer, and drops the outermost level from the index.  Hence me adding it back in on 174.  If you do the same thing on a dataframe with a normal index and the indexer corresponds to one row, it coerces the row to a series.  

In this situation, most of our data is (location_id, date), but we have some dataframes that are just by location.